modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextAction.java

```diff
@@ -0,0 +1,251 @@
+ * <ul>
+ *     <li> GET /_scripts/painless/_context -- retrieves a list of contexts </li>
+ *     <li> GET /_scripts/painless/_context?context=%name% --
+ *     retrieves all available information about the API for this specific context</li>
+ * </ul>
+ */
+public class PainlessContextAction extends Action<PainlessContextAction.Response> {
+
+    public static final PainlessContextAction INSTANCE = new PainlessContextAction();
+    private static final String NAME = "cluster:admin/scripts/painless/context";
+
+    private static final String SCRIPT_CONTEXT_NAME_PARAM = "context";
+
+    private PainlessContextAction() {
+        super(NAME);
+    }
+
+    @Override
+    public Response newResponse() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Writeable.Reader<Response> getResponseReader() {
+        return Response::new;
+    }
+
+    public static class Request extends ActionRequest {
+
+        private String scriptContextName;
+
+        public Request() {
+            scriptContextName = null;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            scriptContextName = in.readString();
+        }
+
+        public void setScriptContextName(String scriptContextName) {
+            this.scriptContextName = scriptContextName;
+        }
+
+        public String getScriptContextName() {
+            return scriptContextName;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(scriptContextName);
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        public static final ParseField CONTEXTS = new ParseField("contexts");
+
+        private final List<String> scriptContextNames;
+        private final PainlessContextInfo painlessContextInfo;
+
+        public Response(List<String> scriptContextNames, PainlessContextInfo painlessContextInfo) {
+            if (scriptContextNames == null) {
+                this.scriptContextNames = null;
+            } else {
+                scriptContextNames = new ArrayList<>(scriptContextNames);
+                scriptContextNames.sort(String::compareTo);
+                this.scriptContextNames = Collections.unmodifiableList(scriptContextNames);
+            }
+
+            this.painlessContextInfo = painlessContextInfo;
+        }
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            scriptContextNames = in.readBoolean() ? in.readStringList() : null;
+            painlessContextInfo = in.readBoolean() ? new PainlessContextInfo(in) : null;
```

Comment:
instead of the ternary expressions maybe use: `readOptionalWriteable(...)` and just use an empty list for scriptContextNames (if there isn't anything)?

Reply:
Thank you for readOptionalWritable!  That's what I was looking for, but failed to find it.